source/common/filesystem/win32/filesystem_impl.cc

```diff
@@ -0,0 +1,100 @@
+#include <fcntl.h>
+#include <io.h>
+#include <sys/stat.h>
+#include <windows.h>
+
+#undef DELETE
+#undef GetMessage
+
+#include "common/common/assert.h"
+#include "common/filesystem/filesystem_impl.h"
+
+#include <fstream>
+#include <iostream>
+#include <sstream>
+#include <string>
+
+#include "envoy/common/exception.h"
+
+#include "common/common/fmt.h"
+
+namespace Envoy {
+namespace Filesystem {
+
+FileImplWin32::~FileImplWin32() {
+  if (isOpen()) {
+    const Api::IoCallBoolResult result = close();
+    ASSERT(result.rc_);
+  }
+}
+
+void FileImplWin32::openFile() {
+  const int flags = _O_RDWR | _O_APPEND | _O_CREAT;
+  const int mode = _S_IREAD | _S_IWRITE;
+
+  fd_ = ::_open(path_.c_str(), flags, mode);
+}
+
+ssize_t FileImplWin32::writeFile(absl::string_view buffer) {
+  if (fd_ == -1) {
+    errno = EBADF;
+    return -1;
+  }
+
+  return ::_write(fd_, buffer.data(), buffer.size());
+}
+
+bool FileImplWin32::closeFile() { return ::_close(fd_) != -1; }
+
+FilePtr InstanceImplWin32::createFile(const std::string& path) {
+  return std::make_unique<FileImplWin32>(path);
+}
+
+bool InstanceImplWin32::fileExists(const std::string& path) {
+  const DWORD attributes = ::GetFileAttributes(path.c_str());
+  return attributes != INVALID_FILE_ATTRIBUTES;
+}
+
+bool InstanceImplWin32::directoryExists(const std::string& path) {
+  const DWORD attributes = ::GetFileAttributes(path.c_str());
+  if (attributes == INVALID_FILE_ATTRIBUTES) {
+    return false;
+  }
+  return attributes & FILE_ATTRIBUTE_DIRECTORY;
+}
+
+ssize_t InstanceImplWin32::fileSize(const std::string& path) {
+  struct _stat info;
+  if (::_stat(path.c_str(), &info) != 0) {
+    return -1;
+  }
+  return info.st_size;
+}
+
+std::string InstanceImplWin32::fileReadToEnd(const std::string& path) {
+  if (illegalPath(path)) {
+    throw EnvoyException(fmt::format("Invalid path: {}", path));
+  }
+
+  std::ios::sync_with_stdio(false);
```

Comment:
Does this change some global state?

Reply:
This was copied from the existing Linux implementation:
https://github.com/envoyproxy/envoy/blob/0c0bb54967e199a9b407ece1c4b006b3faeddeaf/source/common/filesystem/filesystem_impl.cc#L104

Reply:
Looks like this has been there for a long long time! I'm actually not sure what it's meant to be doing here, since the documentation implies it only affects the standard streams like stdin, stdout, and stderr. But nothing in this function seems to be dealing with these streams.

@mattklein123 do you remember why this was added?

Reply:
No clue. Can we remove this in a follow up on both sides?